From 41799cb256853242ce079082cf5b822b708a08b9 Mon Sep 17 00:00:00 2001
From: Oliver Eftevaag <oliver.eftevaag@qt.io>
Date: Wed, 1 Oct 2025 10:10:14 +0200
Subject: [PATCH 293/553] Update contrastPreference correctly on Android

Android has an option called "Color contrast" in the
Settings app, found under "Display and Touch".

Qt has a new property in 6.10 called contrastPreference, which can
either be `NoPreference` or `HighContrast`. The intention is to use
native API's to determine if the user has enabled a setting on the
platform level, to increase the overall contrast on the system.
If that is the case, the contrastPreference property should return
`HighContrast`.

This property is then used by our built-in styles, to add more outlines
to controls and potentially change some of the palette colors that are
in use. It can also be used by custom styles.

We're already detecting if the system is running in high contrast mode
on Windows, macOS, Gnome, and WebAssembly. Android took a bit longer to
implement due to the ambiguity of their API. On Android the user has
the option to set the "Color contrast" to either `Default`, `Medium`
or `High`. For now, the contrastPreference property will only return
`HighContrast` if the "Color contrast" is set to `High` only.

When changing this setting on Android, the Activity is recreated.
It is desireable to handle this configuration change ourselves, however,
we've not been able to find a good approach here. Most configuration
changes can be added to a list in AndroidManifest.xml to override the
default behavior of resetting the Activity, but 'colorMode' or any other
handler doesn't seem to work when changing the contrast setting.

Fixes: QTBUG-139294
Change-Id: Iaebb89c79feb749655132d2db83c1bb1a67b301d
Reviewed-by: Assam Boudjelthia <assam.boudjelthia@qt.io>
(cherry picked from commit 2dfffefa6d408222849ec0afe2df8d8cea205b16)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 .../qt/android/QtActivityDelegateBase.java        |  9 +++++++++
 .../platforms/android/qandroidplatformtheme.cpp   | 15 +++++++++++++++
 .../platforms/android/qandroidplatformtheme.h     |  1 +
 3 files changed, 25 insertions(+)

diff --git a/src/android/jar/src/org/qtproject/qt/android/QtActivityDelegateBase.java b/src/android/jar/src/org/qtproject/qt/android/QtActivityDelegateBase.java
index 4734e439a50..35f519ca9a4 100644
--- a/src/android/jar/src/org/qtproject/qt/android/QtActivityDelegateBase.java
+++ b/src/android/jar/src/org/qtproject/qt/android/QtActivityDelegateBase.java
@@ -6,6 +6,7 @@
 package org.qtproject.qt.android;
 
 import android.app.Activity;
+import android.app.UiModeManager;
 import android.content.pm.ActivityInfo;
 import android.content.pm.PackageManager;
 import android.content.res.Configuration;
@@ -29,6 +30,7 @@ abstract class QtActivityDelegateBase
     private boolean m_contextMenuVisible = false;
 
     static native boolean canOverrideColorSchemeHint();
+    static native void updateUiContrast(float newUiContrast);
 
     // Subclass must implement these
     abstract void startNativeApplicationImpl(String appParams, String mainLib);
@@ -129,5 +131,12 @@ abstract class QtActivityDelegateBase
                 QtDisplayManager.handleUiDarkModeChanged(1);
                 break;
         }
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
+            // FIXME: Handle contrast changes the same way as uiMode changes (QTBUG-140749).
+            UiModeManager uiModeManager =
+                (UiModeManager) m_activity.getSystemService(m_activity.UI_MODE_SERVICE);
+            updateUiContrast(uiModeManager.getContrast());
+        }
     }
 }
diff --git a/src/plugins/platforms/android/qandroidplatformtheme.cpp b/src/plugins/platforms/android/qandroidplatformtheme.cpp
index 5826c83bdb2..9295ea56c1a 100644
--- a/src/plugins/platforms/android/qandroidplatformtheme.cpp
+++ b/src/plugins/platforms/android/qandroidplatformtheme.cpp
@@ -463,6 +463,21 @@ Java_org_qtproject_qt_android_QtActivityDelegateBase_canOverrideColorSchemeHint(
     return QAndroidPlatformTheme::instance()->colorSchemeOverride() == Qt::ColorScheme::Unknown;
 }
 
+static Qt::ContrastPreference s_contrastPreference = Qt::ContrastPreference::NoPreference;
+
+Qt::ContrastPreference QAndroidPlatformTheme::contrastPreference() const
+{
+    return s_contrastPreference;
+}
+
+extern "C" JNIEXPORT void JNICALL
+Java_org_qtproject_qt_android_QtActivityDelegateBase_updateUiContrast(JNIEnv *, jobject,
+                                                                      jfloat newUiContrast)
+{
+    s_contrastPreference = newUiContrast > 0.5f ? Qt::ContrastPreference::HighContrast
+                                                : Qt::ContrastPreference::NoPreference;
+}
+
 static inline int paletteType(QPlatformTheme::Palette type)
 {
     switch (type) {
diff --git a/src/plugins/platforms/android/qandroidplatformtheme.h b/src/plugins/platforms/android/qandroidplatformtheme.h
index 339919acb75..7e822b88c7d 100644
--- a/src/plugins/platforms/android/qandroidplatformtheme.h
+++ b/src/plugins/platforms/android/qandroidplatformtheme.h
@@ -42,6 +42,7 @@ public:
     Qt::ColorScheme colorScheme() const override;
     Qt::ColorScheme colorSchemeOverride() const { return m_colorSchemeOverride; };
     void requestColorScheme(Qt::ColorScheme scheme) override;
+    Qt::ContrastPreference contrastPreference() const override;
 
     const QPalette *palette(Palette type = SystemPalette) const override;
     const QFont *font(Font type = SystemFont) const override;
-- 
2.51.2

